test_util/marathon.py

```diff
@@ -200,11 +200,46 @@ def _pool_for_marathon_app(app_id):
                 return None
 
         try:
-            return _pool_for_marathon_app(app_definition['id'])
+            return _poll_marathon_for_app_deployment(app_definition['id'])
         except retrying.RetryError:
             raise Exception("Application deployment failed - operation was not "
                             "completed in {} seconds.".format(timeout))
 
+    def ensure_deployments_complete(self, timeout=600):
+        """
+        This method ensures that, there are no pending deployments
+
+        :return: True if all deployments are completed within time out. Raises an exception otherwise.
+        """
+
+        @retrying.retry(wait_fixed=1000, stop_max_delay=120 * 1000, retry_on_exception=lambda x: False)
+        def _get_deployments_json():
+            r = self.get('v2/deployments')
+            r.raise_for_status()
+            return r.json()
+
+        def retry_on_assertion_error(exception):
+            return isinstance(exception, AssertionError)
+
+        @retrying.retry(retry_on_exception=retry_on_assertion_error,
+                        stop_max_attempt_number=10,
+                        wait_fixed=timeout * 1000)
+        def ensure_deployment_is_finished():
+            deployments_json = _get_deployments_json()
+            logging.info("v2/deployments json: {deployments_json}".format(deployments_json=deployments_json))
+            assert not deployments_json, "No deployment should be happening."
+
+        try:
+            ensure_deployment_is_finished()
+        except retrying.RetryError:
+            raise Exception("Deployments were not completed within {timeout} seconds".format(timeout=timeout))
+
+    def deploy_app_and_wait_for_deployment(self, app_definition, timeout=120):
```

Comment:
* Why the different default timeout here? 120 vs 600 in `ensure_deployments_complete()`.
* I don't think deploy-and-block makes sense as a separate method. That we can have different deploy timeouts to get mismatched is a smell. What if we want to set the `deploy_app()` timeout as well? What if we want to block on the deploy, _and_ allow for tasks to fail during it? In order to get all our options we'd have to tediously duplicate kwargs and pass them through to one of the two methods we're calling here, and make sure kwarg names for the two methods don't overlap and introduce an ambiguity. If we need to provide deploy-and-block behavior, I think it'd make more sense as an optional `deploy_timeout` kwarg for `deploy_app()` that defaults to `None`. If it's `None` we don't block, otherwise we time out at `deploy_timeout` seconds.